spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/s3/S3AutoConfiguration.java

```diff
@@ -78,8 +81,20 @@ S3ClientBuilder s3ClientBuilder(AwsClientBuilderConfigurer awsClientBuilderConfi
 	@ConditionalOnMissingBean(S3Operations.class)
 	@ConditionalOnBean(S3ObjectConverter.class)
 	S3Template s3Template(S3Client s3Client, S3OutputStreamProvider s3OutputStreamProvider,
-			S3ObjectConverter s3ObjectConverter) {
-		return new S3Template(s3Client, s3OutputStreamProvider, s3ObjectConverter);
+			S3ObjectConverter s3ObjectConverter, S3Presigner s3Presigner) {
+		return new S3Template(s3Client, s3OutputStreamProvider, s3ObjectConverter, s3Presigner);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	S3Presigner s3Presigner(S3Properties properties, AwsProperties awsProperties,
+			AwsCredentialsProvider credentialsProvider, AwsRegionProvider regionProvider) {
+		S3Presigner.Builder builder = S3Presigner.builder().serviceConfiguration(properties.toS3Configuration())
+				.credentialsProvider(credentialsProvider).region(regionProvider.getRegion());
+		Optional.ofNullable(awsProperties.getEndpoint()).ifPresent(builder::endpointOverride);
```

Comment:
endpoint can be either set on global `AwsProperties` (`spring.cloud.aws.endpoint`) or specific to S3 `S3Properties` - `spring.cloud.aws.s3.endpoint`. 

if global endpoint set -> set it on builder
if s3 endpoint set -> set it on builder

S3 endpoint configuration if set has higher precedence than endpoint configured globally.